File: src/main/java/com/example/mod/hook/MinecraftHook.java
package com.example.mod.hook;

import net.weavemc.api.Hook;
import org.jetbrains.annotations.NotNull;
import org.objectweb.asm.Opcodes;
import org.objectweb.asm.Type;
import org.objectweb.asm.tree.ClassNode;
import org.objectweb.asm.tree.MethodInsnNode;

public class MinecraftHook extends Hook {
    public MinecraftHook() {
        super("net/minecraft/client/Minecraft");
    }

    @Override
    public void transform(@NotNull ClassNode classNode, @NotNull AssemblerConfig assemblerConfig) {
        classNode.methods.stream()
                .filter(m -> m.name.equals("startGame"))
                .findFirst()
                .ifPresent(m -> m.instructions.insert(
                        new MethodInsnNode(
                                Opcodes.INVOKESTATIC,
                                Type.getInternalName(MinecraftHook.class),
                                "onStartGame",
                                "()V"
                        )
                ));
    }

    @SuppressWarnings("unused")
    public static void onStartGame() {
        System.out.println("Hook Test");
    }
}

File: src/main/kotlin/com/example/mod/hook/MinecraftHook.kt
package com.example.mod.hook

import net.weavemc.api.Hook
import org.objectweb.asm.Opcodes
import org.objectweb.asm.Type
import org.objectweb.asm.tree.ClassNode
import org.objectweb.asm.tree.MethodInsnNode

class MinecraftHook : Hook("net/minecraft/client/Minecraft") {
    override fun transform(node: ClassNode, cfg: AssemblerConfig) {
        node.methods.firstOrNull { method -> method.name == "startGame" }?.let { method ->
            method.instructions.insert(
                MethodInsnNode(
                    Opcodes.INVOKESTATIC,
                    Type.getInternalName(MinecraftHook::class.java),
                    "onStartGame",
                    "()V"
                )
            )
            println("onStartGame transformed")
        }
    }

    companion object {
        @JvmStatic
        @Suppress("unused")
        fun onStartGame() {
            println("Hook Test")
        }
    }
}
